feat(cms,api-gateway,frontend): integrate account page with backend API and member profile management

ericsen-tsai opened the pull request:
## Summary

This PR integrates the account page UI with the backend API, enabling full member profile management functionality. The changes include adding new fields to the Member model (nickname, contactEmail, avatar), implementing photo upload/delete capabilities, enhancing the GraphQL proxy to handle file uploads, and connecting the frontend account page to real member data from the authentication store.

## Changes

### Backend (CMS)
- **Member Model Enhancement**: Added new fields to Member list (`nickname`, `contactEmail`, `avatar` relationship to Photo)
- **Database Migration**: Created migration script to add new Member fields with proper indexes and foreign key constraints
- **Photo Access Control**: Updated Photo list to allow Member role access for avatar management
- **Development Image Handling**: Added static file serving for `/resized` routes in non-production environments
- **2FA Middleware**: Added exception for `/resized/*` routes in non-production to bypass 2FA checks
- **Photo Resized URLs**: Updated Photo resized image handling to use original filename in development environment

### API Gateway
- **File Upload Support**: Enhanced GraphQL proxy middleware to preserve `multipart/form-data` Content-Type for file uploads

### Frontend
- **API Integration**: 
  - Added photo upload and delete API functions with multipart form data support
  - Implemented member profile update mutation with React Query hooks
  - Created GraphQL queries and mutations for member profile management
- **Account Page Integration**:
  - Replaced mock data with real member data from authentication store
  - Implemented avatar upload functionality with file validation
  - Added form handling for profile updates (name, nickname, contactEmail, avatar)
  - Integrated photo upload/delete workflow (delete old avatar before uploading new one)
- **UI/UX Improvements**:
  - Added proper loading states and error handling
  - Updated form validation to handle optional fields
  - Added default values and placeholder text for empty fields
  - Improved date formatting with better validation
- **Next.js Configuration**: Added remote image patterns for `kids-storage.twreporter.org` and `localhost` to support avatar images
- **Constants**: Added `NO_DATA_TEXT` and `DEFAULT_TEXT_HOLDER` constants for consistent empty state display

### Schema Updates
- Updated GraphQL schema to include new Member fields in queries, mutations, and filters
- Updated Prisma schema with new Member fields and Photo relationship

## Additional Notes

- The avatar upload workflow automatically deletes the old avatar photo before uploading a new one to prevent orphaned files
- Development environment uses original filenames for resized images instead of the production naming convention (`filename-size.webp`)
- The account page now properly handles empty/null values with fallback to default text
- All form fields are properly validated with appropriate error messages
- The implementation uses React Query for efficient data fetching and mutation handling
- This PR is based on #835 

## Screenshot(s)

https://github.com/user-attachments/assets/db922edd-f373-41c1-abd6-5c372d30698d



## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/2ae8dbdca9328090847fd53facab35f7?source=copy_link)

ericsen-tsai committed:
feat(cms): enhance Member model with additional fields

- Added 'nickname' and 'contactEmail' fields to the Member model.
- Updated GraphQL schema and input types to include new fields.
- Introduced avatar relationship for Member to Photo model.
- Modified member list to support new fields in the CMS.

ericsen-tsai committed:
refactor(cms): update Photo list configuration and role permissions

- Changed import from ListConfig to ListType for the Photo list.
- Exported the Photo list configuration directly.
- Added RoleEnum.Member to the create permissions for the Photo list.

ericsen-tsai committed:
feat(api-gateway): enhance GraphQL proxy for file uploads

- Preserved original Content-Type for multipart/form-data requests.
- Set Content-Type to application/json for regular GraphQL requests.

ericsen-tsai committed:
feat(api): add photo upload and delete functionality and adjust member profile update

- Implemented `uploadPhoto` and `deletePhoto` functions in the photo API.
- Added GraphQL mutations for creating and deleting photos.
- Introduced `updateMemberProfile` function to update member details with new fields.
- Updated GraphQL schema to include new member profile fields and mutations.
- Created React Query hooks for managing photo uploads and deletions, as well as member profile updates.

ericsen-tsai committed:
fix(utils): improve date validation in getFormattedDate function

- Enhanced the date validation logic to check for invalid date objects using isNaN.
- Ensured that the function returns an empty string for invalid dates.

ericsen-tsai committed:
feat(frontend): integrate membership account management with new fields and image handling

- Added support for remote images from 'kids-storage.twreporter.org' in Next.js configuration.
- Introduced new constants for input fields, including default text for empty data.
- Updated account form schema to include optional fields for nickname and contact email, and modified avatar handling.
- Refactored account management components to utilize new member data structure and improve user experience with default text placeholders.
- Enhanced user avatar component to handle default avatar scenarios.
- Added membership account page integration.

ericsen-tsai committed:
chore(cms): add migration for member avatar list

ericsen-tsai committed:
feat(cms): add resized image handling for development

- Introduced static serving of resized images in non-production environments.
- Updated two-factor authentication middleware to skip 2FA for resized image routes in development.
- Adjusted photo list configuration to conditionally format resized image URLs based on the environment.

nickhsine commented on packages/cms/lists/photo.ts:
```diff
@@ -122,11 +128,10 @@ const listConfigurations: ListConfig<any> = list({
         RoleEnum.Admin,
         RoleEnum.Editor,
         RoleEnum.Contributor,
+        RoleEnum.Member,
```
question:
Should we also grant Member role the permission to delete photo?

ericsen-tsai replied:
Fixed in [fix(frontend): allow member to delete avatar](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/9cc7a4355bae5f0dcc91449afa60925bd938aca8)

nickhsine replied:
question:
Do we need to prevent clients from deleting others' photos?

thought:
Since the photos uploaded by members are different from the ones uploaded by editors, we could add a new list, like member-upload-photo, to handle the custom requirements.  

Custom requirements are like:
1. file size limitation is 5 Mb
2. we don't need to resize the photos into 5 dimensions width.
3. we need to add deletion guard to prevent clients from deleting others' photos.

ericsen-tsai replied:
Adjusted in [refactor(frontend): change photo to member avatar](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/46e2cca09329150979ffef78b8ef129ffbf73da5)

ericsen-tsai committed:
fix(frontend): allow member to delete avatar

ericsen-tsai committed:
fix(frontend): update member profile handling and improve avatar upload logic

- Fixed `updateMemberProfile` to use correct GraphQL mutation types.
- Adjusted avatar URL handling to remove unnecessary fallback.
- Improved avatar upload logic to ensure old avatars are deleted after new uploads.

ericsen-tsai commented on packages/frontend/src/api/member.ts:
```diff
@@ -58,3 +62,28 @@ export const getMemberProfileByMemberId = async ({
+  accessToken: string
+  data: MemberUpdateInput
+}) => {
+  const response = await sendGQLRequest<MemberUpdateInput>(
```
Fixed in [fix(frontend): update member profile handling and improve avatar upload logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/569b6056cee54f46582215b1c4c6cd5bf27876bc)

ericsen-tsai commented on packages/frontend/src/modules/membership/account/index.tsx:
```diff
@@ -0,0 +1,244 @@
+          })
+          avatarId = newAvatar?.id
+          avatarFileRef.current = null
+        }
```
Fixed in [fix(frontend): update member profile handling and improve avatar upload logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/569b6056cee54f46582215b1c4c6cd5bf27876bc)

ericsen-tsai commented on packages/frontend/src/modules/membership/types.ts:
```diff
@@ -1,8 +1,18 @@
+  name: z.string().min(1, '請輸入全名'),
+  nickname: z.string().optional(),
+  contactEmail: z.union([z.email('請輸入有效的電子郵件格式'), z.literal('')]),
+  avatarUrl: z.url().optional(),
```
avatarUrl will not be an empty string in the current scenario, I'll leave it as it is

ericsen-tsai commented on packages/frontend/src/modules/membership/types.ts:
```diff
@@ -1,8 +1,18 @@
+export const accountFormSchema = z.object({
+  name: z.string().min(1, '請輸入全名'),
+  nickname: z.string().optional(),
+  contactEmail: z.union([z.email('請輸入有效的電子郵件格式'), z.literal('')]),
```
in zod v4, email is valid.

https://zod.dev/api?id=string-formats

ericsen-tsai commented on packages/routing-ui/src/components/input.tsx:
```diff
@@ -38,11 +38,23 @@ const inputVariants = cva(
+        error: 'border-semantic-danger hover:border-semantic-danger',
+      },
+      mode: {
+        default: 'rounded-[12px] border-neutral-400',
```
Fixed in [fix(frontend): update member profile handling and improve avatar upload logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/569b6056cee54f46582215b1c4c6cd5bf27876bc)

ericsen-tsai commented on packages/cms/express-mini-apps/two-factor-auth/middleware.ts:
```diff
@@ -167,6 +169,10 @@ export function twoFactorAuthMiddleware(
+      // Also skip 2FA for /resized/* routes in non-production
+      if (envVar.nodeEnv !== 'production' && req.path.startsWith('/resized/')) {
+        return next('route')
+      }
```
Fixed in [fix(frontend): update member profile handling and improve avatar upload logic](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/569b6056cee54f46582215b1c4c6cd5bf27876bc)

ericsen-tsai committed:
chore(routing-ui): bump version to 0.1.1

ericsen-tsai committed:
feat(cms): introduce MemberAvatar model and update member avatar handling

- Added new MemberAvatar model to manage member avatars with fields for name, image file, and relationships.
- Updated Member model to reference MemberAvatar instead of Photo for avatar handling.
- Enhanced GraphQL schema with mutations and queries for creating, updating, and deleting member avatars.
- Adjusted access control for member avatars to allow specific roles to manage avatars.

ericsen-tsai committed:
refactor(cms): remove non-production 2FA skip for resized image routes

- Eliminated the conditional logic that skipped two-factor authentication for resized image routes in non-production environments.

ericsen-tsai committed:
refactor(frontend): change photo to member avatar

- Added new API functions for uploading and deleting member avatars, including GraphQL mutations for handling these operations.
- Created React Query hooks to facilitate avatar upload and deletion in the membership account management.
- Updated account management components to utilize the new avatar handling logic, ensuring seamless integration with the member profile.
- Removed photo related API and React Query hooks.

ericsen-tsai committed:
fix(frontend): enhance input component style

nickhsine commented on packages/cms/lists/member-avatar.ts:
```diff
@@ -0,0 +1,148 @@
+        return sessionMemberId === memberId
+      },
+    },
+  },
```
thought:
可以參考 https://github.com/kids-reporter/kids-reporter-monorepo/blob/ndx/packages/cms/lists/post-choice-answer.ts#L132-L136 ，使用 filterAccessControl 來阻擋使用者刪除或是修改非自己的檔案。

access.item 也可以達到一樣的目的，但 item acl 好像更適合複雜的 acl 設定，如果單純是使用者僅能修刪改自己的檔案，好像 filter 滿適合的？

ericsen-tsai replied:
Fixed in [fix(cms): streamline access control for member avatars](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/77103d6559e208112c52f5c19cadb211b4a2e9ab)

nickhsine commented on packages/cms/lists/member-avatar.ts:
```diff
@@ -0,0 +1,148 @@
+        RoleEnum.Admin,
+        RoleEnum.Editor,
+        RoleEnum.Member,
+      ]),
```
thought:
關於使用者產生的資料 (user generated content)，是不是可以統一權限設定，例如：https://github.com/kids-reporter/kids-reporter-monorepo/blob/ndx/packages/cms/lists/post-choice-answer.ts#L126-L131

ericsen-tsai replied:
Fixed in [fix(cms): streamline access control for member avatars](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/77103d6559e208112c52f5c19cadb211b4a2e9ab)

nickhsine reviewed (commented):
suggestion:
db migration script 需要更換成 MemberAvatar 的版本。

nickhsine commented:
Note:
目前 frontend 是使用 MemberAvatar.fileUrl，該檔案是原始檔案，我們之後可能會需要使用 resize 過的版本。

ericsen-tsai committed:
fix(cms): streamline access control for member avatars

- Reorganized access control logic for member avatars by introducing dedicated functions for operation access and filtering.
- Simplified role-based permissions for create, update, and delete operations, ensuring consistent access management.
- Removed redundant deletion logic, enhancing clarity and maintainability of the code.

ericsen-tsai commented:
> suggestion: db migration script 需要更換成 MemberAvatar 的版本。

Added in [chore(cms): add migration for member avatar list](https://github.com/kids-reporter/kids-reporter-monorepo/pull/837/commits/6a476a1410f27324001edf8fecf7021dd97ceb61)

nickhsine commented:
LGTM 👍 👍 👍 
Conflicts 再麻煩解決了
